File: MPI_Code/Fortran/struct.f90
PROGRAM type_struct
USE mpi_f08
IMPLICIT NONE
INTEGER::rank
TYPE ddt
  INTEGER::num
  real::x
  COMPLEX::data(4)
END TYPE ddt
INTEGER::blocklengths(3)
INTEGER(KIND=MPI_ADDRESS_KIND)::disps(3)
TYPE(MPI_Datatype)::arr_types(3),restype
TYPE(ddt)::a
INTEGER::i,size
blocklengths=(/1,1,4/)
arr_types=(/MPI_INTEGER,MPI_REAL,MPI_COMPLEX/)
CALL MPI_Init
CALL MPI_Comm_rank(MPI_COMM_WORLD,rank)
CALL MPI_Get_address(a%num,disps(1))
CALL MPI_Get_address(a%x,disps(2))
CALL MPI_Get_address(a%data,disps(3))
disps(3)=disps(3)-disps(1);   disps(2)=disps(2)-disps(1);  disps(1)=MPI_BOTTOM
CALL MPI_Type_create_struct(3,blocklengths,disps,arr_types,restype)
CALL MPI_Type_commit(restype)
IF(rank==0)THEN
  a%num=6;  a%x=3.14
  DO i=1,4
    a%data(i)=cmplx(i,i)
  END DO
  CALL MPI_Send(a,1,restype, 1, 30, MPI_COMM_WORLD)
ELSEIF(rank==1)THEN
  CALL MPI_Recv(a,1,restype, 0, 30, MPI_COMM_WORLD,MPI_STATUS_IGNORE)
  PRINT*,'P:',rank,' I got'
  PRINT*,a%num
  PRINT*,a%x
  PRINT*,a%data
ENDIF
CALL MPI_Type_free(restype)
CALL MPI_Finalize
END PROGRAM type_struct

File: example_code/mpi/MPI_Code/C/struct.c
#include <stdio.h>
#include "mpi.h"
int main()
{
  int rank,i;
  struct{ int num;  float x; double data[4];} a;
  int blocklengths[3]={1,1,4};
  MPI_Datatype types[3]={MPI_INT, MPI_FLOAT, MPI_DOUBLE};
  MPI_Aint disps[3];
  MPI_Datatype restype;
  
  MPI_Init(NULL,NULL); 
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);
  MPI_Get_address(&a.num,&disps[0]);
  MPI_Get_address(&a.x,&disps[1]);
  MPI_Get_address(&a.data,&disps[2]);
  disps[2]-=disps[0];   disps[1]-=disps[0];   disps[0] = (MPI_Aint)MPI_BOTTOM;
  MPI_Type_create_struct(3,blocklengths,disps,types,&restype);
  MPI_Type_commit(&restype);
  if(rank==0){
    a.num=6,   a.x=3.14;
    for(i=0;i<4;i++) a.data[i]=(double)i;
    MPI_Send(&a,1,restype,1,30,MPI_COMM_WORLD);
  }else if(rank==1){
    MPI_Recv(&a,1,restype,0,30,MPI_COMM_WORLD,MPI_STATUS_IGNORE);
    printf("P: %d my a is %d %f %lf %lf %lf %lf\n", rank,a.num, a.x,\
           a.data[0],a.data[1], a.data[2], a.data[3]);
  }
  MPI_Type_free(&restype);
  MPI_Finalize();
  return 0;
}
